heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/MesosScheduler.java

```diff
@@ -0,0 +1,232 @@
+import com.twitter.heron.spi.scheduler.IScheduler;
+import com.twitter.heron.spi.utils.Runtime;
+import com.twitter.heron.spi.utils.SchedulerUtils;
+
+/**
+ * Schedule a topology to a mesos cluster
+ */
+
+public class MesosScheduler implements IScheduler {
+  private static final Logger LOG = Logger.getLogger(MesosScheduler.class.getName());
+
+  private Config config;
+  private Config runtime;
+  private MesosFramework mesosFramework;
+  private SchedulerDriver driver;
+
+
+  @Override
+  public void initialize(Config mConfig, Config mRuntime) {
+    this.config = mConfig;
+    this.runtime = mRuntime;
+    this.mesosFramework = getMesosFramework();
+
+  }
+
+  /**
+   * Waits for the driver to be stopped or aborted
+   *
+   * @param timeout maximum time waiting
+   * @param unit time unit to wait
+   */
+  protected void joinSchedulerDriver(long timeout, TimeUnit unit) {
+    // ExecutorService used to monitor whether close() completes in time
+    ExecutorService service = Executors.newFixedThreadPool(1);
+    final CountDownLatch closeLatch = new CountDownLatch(1);
+    Runnable driverJoin = new Runnable() {
+      @Override
+      public void run() {
+        driver.join();
+        closeLatch.countDown();
+      }
+    };
+    service.submit(driverJoin);
+
+    LOG.info("Waiting for Mesos Driver got stopped");
+    try {
+      if (!closeLatch.await(timeout, unit)) {
+        LOG.severe("Mesos Driver failed to stop in time.");
+      } else {
+        LOG.info("Mesos Driver stopped.");
+      }
+    } catch (InterruptedException e) {
+      LOG.log(Level.SEVERE, "Close latch thread is interrupted: ", e);
+    }
+
+    // Shutdown the ExecutorService
+    service.shutdownNow();
+  }
+
+  @Override
+  public void close() {
+    // Stop the SchedulerDriver
+    if (driver != null) {
+      // Stop the SchedulerDriver
+      if (this.mesosFramework.isTerminated()) {
+        driver.stop();
+      } else {
+        driver.stop(true);
+      }
+
+      // Waits for the driver to be stopped or aborted,
+      // i.e. waits for the message sent to Mesos Master
+      long stopTimeoutInMs = MesosContext.getHeronMesosSchedulerDriverStopTimeoutMs(config);
+      joinSchedulerDriver(stopTimeoutInMs, TimeUnit.MILLISECONDS);
+    }
+
+    // Will not kill the topology when close() is invoked,
+    // since the lifecycle of Topology is independent from Scheduler
+  }
+
+  @Override
+  public boolean onSchedule(PackingPlan packing) {
+    // Construct the jobDefinition
+    Map<Integer, BaseContainer> jobDefinition = new HashMap<>();
+
+    for (int containerIndex = 0;
+         containerIndex < Runtime.numContainers(runtime);
+         containerIndex++) {
+      jobDefinition.put(containerIndex, getBaseContainer(containerIndex, packing));
+    }
+
+    String masterURI = MesosContext.getHeronMesosMasterUri(config);
+    this.driver = getSchedulerDriver(masterURI, mesosFramework);
+
+    // start the driver, non-blocking
+    driver.start();
```

Comment:
Why we start the driver inside onSchedule() instead of inside initialize()? Put it in initialize seems more reasonable  and all the information needed is also available in initialize()